feat: documentation for PSS (#23)
ethersphere/bee-js-docs: docs/user-documentation/pss.md

---
title: Postal Service over Swarm
id: pss
slug: /pss
sidebar_label: Postal Service over Swarm
---

import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

> Swarm provides the ability to send messages that appear to be normal Swarm traffic, but are in fact messages that may be received and decrypted to reveal their content only by the specific nodes they were intended to be received by.

Reviewer: Suggestion: `... only to specific nodes that were intended to receive them`

Author: ```suggestion
Swarm provides the ability to send messages that appear to be normal Swarm traffic, but are in fact messages that may be received and decrypted to reveal their content only to specific nodes that were intended to receive them.
```

PSS provides a pub-sub facility that can be used for a variety of tasks. Nodes are able to listen to messages received for a specific topic in their nearest neighbourhood and create messages destined for another neighbourhood which are sent over the network using Swarm's usual data dissemination protocols.

> The intended use of PSS is to send one of encrypted or public communication with a publicly known identity (to for example initiate further communication directly). Due to the cost of mining the trojan chunks, it is not recommended to use as and instant messaging system. For the same reason, the API is disabled on public gateways.

Reviewer: Suggestion: 

> The intended use of PSS is **to communicate privately** with a publicly known identity (to for example initiate further communication directly). Due to the cost of mining the trojan chunks, it is not recommended to use as **an** instant messaging system.

Author: ```suggestion
The intended use of PSS is to communicate privately with a publicly known identity (to for example initiate further communication directly). Due to the cost of mining the trojan chunks, it is not recommended to use as an instant messaging system.
```

- The **Swarm network address** defines your location in the kademlia and within the context of PSS is used for addressing the trojan chunks to you. In other words, others may use it to send you a message.
- **PSS public key** can be used by others to encrypt their messages for you.

<!---
### Deriving swarm address from ethereum address
This section will need a lot of love and testing, probably should be in some advanced page but leaving it here as comment since we want to write it at some point.
-->

## Sending message

To send data simply define a topic, prefix of the recipient's swarm network address (we recommend 4-6 character prefix length) and the data to be send.

<Tabs
  groupId="lang_preferrence"
  defaultValue="ts"
  values={[
    {label: 'TypeScript', value: 'ts'},
    {label: 'JavaScript', value: 'js'},
  ]}>
  <TabItem value="ts">

```ts
/**
 * @param {string}            topic
 * @param {string}            targetPrefix
 * @param {string|Uint8Array} data
 * @param {string}            encryptionKey
 */

> bee.pssSend('topic', '9e2e', 'Hello!')

Reviewer: `encryptionKey` is missing from the example. I think it would be good to explain what it does and what happens if you omit it.

Author: It is not missing in the example, it is deliberately omitted here. That being said I should (and will) explain what happens as you said.

```

  </TabItem>
  <TabItem value="js">

```js
/**
 * @param {string}            topic
 * @param {string}            targetPrefix
 * @param {string|Uint8Array} data
 * @param {string}            encryptionKey
 */
bee.pssSend('topic', '9e2e', 'Hello!')
```

  </TabItem>
</Tabs>

Optionally, if you want to encrypt the message, you may provide the recipient's PSS public key.

<Tabs
  groupId="lang_preferrence"
  defaultValue="ts"
  values={[
    {label: 'TypeScript', value: 'ts'},
    {label: 'JavaScript', value: 'js'},
  ]}>
  <TabItem value="ts">

```ts
bee.pssSend(
  'topic',
  '9e2e',
  'Encrypted Hello!',
  '02fa24cac43531176d21678900b37bd800c93da3b02c5e11572fb6a96ec49527fa',
)
```

  </TabItem>
  <TabItem value="js">

```js
bee.pssSend(
  'topic',
  '9e2e',
  'Encrypted Hello!',
  '02fa24cac43531176d21678900b37bd800c93da3b02c5e11572fb6a96ec49527fa',
)
```

  </TabItem>
</Tabs>

## Retrieving message

> As a recipient, you have two ways how to recieve the message. If you are expecting one off message (which is the intended PSS usecase to exchange credentials for further direct communication), you can use the `pssReceive` method.

Reviewer 2: ```suggestion
As a recipient, you have two ways how to receive the message. If you are expecting one off message (which is the intended PSS use case to exchange credentials for further direct communication), you can use the `pssReceive` method.
```
